places/src/api/matcher.rs

```diff
@@ -0,0 +1,303 @@
+    /// A list of reasons why this matched.
+    pub reasons: Vec<MatchReason>,
+}
+
+impl Match {
+    /// Default search behaviors from Desktop: HISTORY, BOOKMARK, OPENPAGE, SEARCHES.
+    /// Default match behavior: MATCH_BOUNDARY_ANYWHERE.
+    pub fn from_adaptive_row(row: &rusqlite::Row) -> rusqlite::Result<Self> {
+        let mut reasons = Vec::new();
+
+        let place_id = row.get_checked::<_, i64>("id")?;
+        let url = row.get_checked::<_, String>("url")?;
+        let history_title = row.get_checked::<_, Option<String>>("title")?;
+        let bookmarked = row.get_checked::<_, String>("bookmarked")?;
+        let bookmark_title = row.get_checked::<_, Option<String>>("btitle")?;
+        let tags = row.get_checked::<_, Option<String>>("tags")?;
+        let frecency = row.get_checked::<_, i64>("frecency")?;
+
+        let title = bookmark_title.or_else(|| history_title).unwrap_or_default();
+
+        let label = if let Some(tags) = tags {
+            reasons.push(MatchReason::Tags(tags));
+            MatchLabel::BookmarkTag
+        } else {
+            MatchLabel::Bookmark
+        };
+        let url = Url::parse(&url).expect("Invalid URL in Places");
+
+        Ok(Self {
+            url,
+            title,
+            label,
+            icon_url: None,
+            frecency,
+            reasons,
+        })
+    }
+}
+
+pub enum MatchBehavior {
+    Anywhere = 0,
+    BoundaryAnywhere = 1,
+}
+
+impl FromSql for MatchBehavior {
+    fn column_result(value: ValueRef) -> FromSqlResult<Self> {
+        Ok(match value.as_i64()? {
+            0 => MatchBehavior::Anywhere,
+            1 => MatchBehavior::BoundaryAnywhere,
+            _ => Err(FromSqlError::InvalidType)?,
+        })
+    }
+}
+
+impl ToSql for MatchBehavior {
+    fn to_sql(&self) -> rusqlite::Result<ToSqlOutput> {
+        Ok(match self {
+            MatchBehavior::Anywhere => ToSqlOutput::from(0i64),
+            MatchBehavior::BoundaryAnywhere => ToSqlOutput::from(1i64),
+        })
+    }
+}
+
+struct Adaptive<'query, 'conn> {
+    query: &'query str,
+    conn: &'conn PlacesDb,
+    max_results: usize,
+    match_behavior: MatchBehavior,
+}
+
+impl<'query, 'conn> Adaptive<'query, 'conn> {
+    pub fn new(
+        query: &'query str,
+        conn: &'conn PlacesDb,
+        max_results: usize,
+    ) -> Adaptive<'query, 'conn> {
+        Adaptive::with_behavior(query, conn, max_results, MatchBehavior::BoundaryAnywhere)
+    }
+
+    pub fn with_behavior(
+        query: &'query str,
+        conn: &'conn PlacesDb,
+        max_results: usize,
+        match_behavior: MatchBehavior,
+    ) -> Adaptive<'query, 'conn> {
+        Adaptive {
+            query,
+            conn,
+            max_results,
+            match_behavior,
+        }
+    }
+
+    pub fn search(&self) -> Result<Vec<Match>> {
+        let mut stmt = self.conn.db.prepare(
+            "
+            SELECT h.url, h.title,
+                   EXISTS(SELECT 1 FROM moz_bookmarks
```

Comment:
probably comment this out/replace with `0 AS bookmarked` while moz_bookmarks is still missing. (ditto for `btitle`), and for the query below.